### PR TITLE
Timer observable

### PR DESCRIPTION
**Pull request by davidwdan**

### CODE REVIEW

File: lib/Rx/Observable/TimerObservable.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                    $observer->onNext($counter++);
+                },
+                $this->dueTime,
+                $this->period
+            );
+        } else {
+            $disposable = $scheduler->schedule(
+                function () use (&$counter, $observer) {
+                    $observer->onNext($counter);
+                    $observer->onCompleted();
```
Comment (mbonneau) [inline lib/Rx/Observable/TimerObservable.php]: Should complete be scheduled separately?

Comment (davidwdan) [inline lib/Rx/Observable/TimerObservable.php]: You would think, but that's the way that RxJS works...

File: test/Rx/Functional/Observable/TimerObservableTest.php [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+                onNext(500, 0),
+                onNext(600, 1),
+                onNext(700, 2),
+                onNext(800, 3),
+                onNext(900, 4)
+            ],
+            $results->getMessages()
+        );
+    }
+}
```
Comment (mbonneau) [inline test/Rx/Functional/Observable/TimerObservableTest.php:129]: Could you add a test that disposes during the subscription period?

Comment (davidwdan) [inline test/Rx/Functional/Observable/TimerObservableTest.php:129]: @mbonneau added some more dispose tests

File: lib/Rx/Observable.php [hunk 3]
```diff
@@ -504,6 +505,19 @@ public function skipUntil(ObservableInterface $other)
...
 
     /**
+     * Returns an observable sequence that produces a value after dueTime has elapsed and then after each period.
+     *
+     * @param $dueTime
+     * @param int|null $period
+     * @param SchedulerInterface $scheduler
+     * @return TimerObservable
+     */
+    public static function timer($dueTime, $period = 0, SchedulerInterface $scheduler = null)
```
Comment (asm89) [inline lib/Rx/Observable.php]: Thinking about the API here. What we currently have seems more like `Interval` to me. [1](http://reactivex.io/documentation/operators/interval.html) `Timer` [2](http://reactivex.io/documentation/operators/timer.html) is emitting one item after an interval, though RxJS has a version that seems to be a combination of both (e.g. an interval that starts _after_ the timer has fired).

Let's add two operators, one `Timer` and one `Interval` that behaves like described and then add some helper methods? E.g.

```
timer($dueTime, $scheduler);
interval($period, $scheduler);
intervalStartingAfter($dueTime, $period, $scheduler); // Can be implemented in terms of the other two?
```

File: lib/Rx/Observable/TimerObservable.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
+<?php
+
+namespace Rx\Observable;
+
+use Rx\Observable;
+use Rx\ObserverInterface;
+use Rx\SchedulerInterface;
+
+class TimerObservable extends Observable
```
Comment (asm89) [inline lib/Rx/Observable/TimerObservable.php:9]: As mentioned above, let's split this in a `Timer` and `Interval`. Should simplify the code a tiny bit. E.g. the `$period > 0` check in `subscribe()` basically shows the that this class has two separate behaviors.

File: lib/Rx/Observable/TimerObservable.php (hunk 1 above)
Comment (asm89) [inline lib/Rx/Observable/TimerObservable.php]: I think we do this in other places as well. We can revisit it later. Good question though.

File: lib/Rx/Observable/TimerObservable.php (hunk 4 above)
Comment (asm89) [inline lib/Rx/Observable/TimerObservable.php:9]: Oh, we already have `Interval` with the delay. So we should probably just slim this one down then. :) 

https://github.com/ReactiveX/RxPHP/blob/master/lib/Rx/Observable/IntervalObservable.php

Comment (davidwdan) [inline lib/Rx/Observable/TimerObservable.php:9]: The other option would be:

```
timer($dueTime, $scheduler);
timerPeriodic($dueTime, $period, $scheduler)
```

And split the observable classes likewise.  

We can accomplish `timerPeriodic` with `interval` and `delay`, so it's probably not necessary.  We can always add it later, if we need to.

Comment (asm89) [inline lib/Rx/Observable/TimerObservable.php:9]: Yeah, let's revisit adding other convenience helpers in a separate issue/PR.

File: test/Rx/Functional/Observable/TimerObservableTest.php [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+     * @test
+     * @expectedException \InvalidArgumentException
+     */
+    public function timer_one_shot_relative_time_zero_non_int()
+    {
+        $results = $this->scheduler->startWithCreate(function () {
+            return Observable::timer('1', $this->scheduler);
+        });
+
+        $this->assertMessages(
```
Comment (asm89) [inline test/Rx/Functional/Observable/TimerObservableTest.php]: unneeded assert, will never be reached?

**Pull request merged**